File: golang/datastructure/tree/avl/avl.go
// 设计平衡二叉搜索树 https://zh.wikipedia.org/wiki/AVL%E6%A0%91
package main

import "fmt"

type TreeNode struct {
	val    int
	height int
	Left   *TreeNode
	Right  *TreeNode
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func GetHeight(root *TreeNode) int {
	if root == nil {
		return 0
	}
	return root.height
}

func Find(root *TreeNode, val int) *TreeNode {
	if root == nil {
		return nil
	}
	if root.val == val {
		return root
	} else if val < root.val {
		return Find(root.Left, val)
	} else {
		return Find(root.Right, val)
	}
}

func findSmallerMax(root *TreeNode) *TreeNode {
	t := root.Left
	for t.Right != nil {
		t = t.Right
	}
	return t
}

func Add(root *TreeNode, val int) *TreeNode {
	if root == nil {
		root = &TreeNode{val: val}
	} else if root.val == val {
		fmt.Println("Val was already in the tree.")
		return root
	} else if val < root.val {
		root.Left = Add(root.Left, val)
		if GetHeight(root.Left)-GetHeight(root.Right) == 2 {
			if val < root.Left.val {
				root = rotateRight(root, root.Left)
			} else {
				root = rotateLeftRight(root, root.Left)
			}
		}
	} else {
		root.Right = Add(root.Right, val)
		if GetHeight(root.Right)-GetHeight(root.Left) == 2 {
			if val > root.Right.val {
				root = rotateLeft(root, root.Right)
			} else {
				root = rotateRightLeft(root, root.Right)
			}
		}
	}
	adjHeight(root)
	return root
}

func Delete(root *TreeNode, val int) *TreeNode {
	if root == nil {
		fmt.Println("Empty tree.")
		return nil
	} else if root.val == val {
		if root.Left == nil {
			root = root.Right
		} else if root.Right == nil {
			root = root.Left
		} else {
			sm := findSmallerMax(root)
			root.val = sm.val
			root.Left = Delete(root.Left, sm.val)
			if GetHeight(root.Right)-GetHeight(root.Left) == 2 {
				if GetHeight(root.Right.Right) >= GetHeight(root.Right.Left) {
					root = rotateLeft(root, root.Right)
				} else {
					root = rotateRightLeft(root, root.Right)
				}
			}
		}
	} else if val < root.val {
		root.Left = Delete(root.Left, val)
		if GetHeight(root.Right)-GetHeight(root.Left) == 2 {
			if GetHeight(root.Right.Right) >= GetHeight(root.Right.Left) {
				root = rotateLeft(root, root.Right)
			} else {
				root = rotateRightLeft(root, root.Right)
			}
		}
	} else {
		root.Right = Delete(root.Right, val)
		if GetHeight(root.Left)-GetHeight(root.Right) == 2 {
			if GetHeight(root.Left.Left) >= GetHeight(root.Left.Right) {
				root = rotateRight(root, root.Left)
			} else {
				root = rotateLeftRight(root, root.Left)
			}
		}
	}
	if root != nil {
		adjHeight(root)
	}
	return root
}

func adjHeight(nodes ...*TreeNode) {
	for _, t := range nodes {
		leftHeight, rightHeight := GetHeight(t.Left), GetHeight(t.Right)
		t.height = max(leftHeight, rightHeight) + 1
	}
}

func rotateLeft(root *TreeNode, node *TreeNode) *TreeNode {
	root.Right = node.Left
	node.Left = root
	adjHeight(root, node)
	return node
}

func rotateRight(root *TreeNode, node *TreeNode) *TreeNode {
	root.Left = node.Right
	node.Right = root
	adjHeight(root, node)
	return node
}

func rotateLeftRight(root *TreeNode, node *TreeNode) *TreeNode {
	root.Left = rotateLeft(node, node.Right)
	return rotateRight(root, root.Left)
}

func rotateRightLeft(root *TreeNode, node *TreeNode) *TreeNode {
	root.Right = rotateRight(node, node.Left)
	return rotateLeft(root, root.Right)
}

func PrintTree(root *TreeNode, t, level int) {
	if root == nil {
		return
	}
	PrintTree(root.Right, 2, level+1)
	switch t {
	case 0:
		fmt.Printf("%2d\n", root.val)
	case 1:
		for i := 0; i < level; i++ {
			fmt.Print("\t")
		}
		fmt.Printf("\\ %2d\n", root.val)
	case 2:
		for i := 0; i < level; i++ {
			fmt.Print("\t")
		}
		fmt.Printf("/ %2d\n", root.val)
	}
	PrintTree(root.Left, 1, level+1)
}

func main() {
	fmt.Println("Test AVL Tree:")
	fmt.Println("Commands:")
	fmt.Println("[0] exit")
	fmt.Println("[1] add value into the tree.")
	fmt.Println("[2] delete value from the tree.")
	fmt.Println("[3] get the height from the tree.")

	cmd := -1
	var tree *TreeNode
	for cmd != 0 {
		_, err := fmt.Scanf("%d", &cmd)
		if err != nil {
			break
		}
		switch cmd {
		case 1:
			var val int
			_, err := fmt.Scanf("%d", &val)
			if err != nil {
				break
			}
			tree = Add(tree, val)
			PrintTree(tree, 0, 0)
		case 2:
			var val int
			_, err := fmt.Scanf("%d", &val)
			if err != nil {
				break
			}
			tree = Delete(tree, val)
			PrintTree(tree, 0, 0)
		case 3:
			var val int
			_, err := fmt.Scanf("%d", &val)
			if err != nil {
				break
			}
			t := Find(tree, val)
			fmt.Println(GetHeight(t))
		}
	}
}

File: c/datastructure/tree/avl/avl.c
// 2. 设计平衡二叉搜索树 https://zh.wikipedia.org/wiki/AVL%E6%A0%91
#include "avl.h"
#include <stdio.h>
#include <stdlib.h>

Tree create_tree(void) {
    return NULL;
}

Tree make_empty(Tree root) {
    if(root != NULL) {
        make_empty(root->left);
        make_empty(root->right);
        free(root);
    }
    return NULL;
}

PtrToNode find(Tree root, int value) {
    if (root == NULL) {
        return NULL;
    } else if (root->value == value) {
        return root;
    } else if (root->value > value) {
        return find(root->left, value);
    } else {
        return find(root->right, value);
    }
}

int tree_height(Tree root) {
    if (root == NULL) {
        return 0;
    }
    return root->height;
}

Tree insert(Tree root, int value) {
    if (root == NULL) {
        root = (Tree) malloc(sizeof(Node));
        if (root == NULL) {
            printf("Not enough space.\n");
            return NULL;
        }
        root->value = value;
        root->left = NULL;
        root->right = NULL;
    } else if (root->value == value) {
        printf("Value was already in the tree.\n");
    } else if (root->value > value) {
        root->left = insert(root->left, value);
        if(tree_height(root->left) - tree_height(root->right) == 2) {
            if (value < root->left->value) {
                root = rotate_right(root, root->left);
            } else {
                root = rotate_left_right(root, root->left);
            }
        }
    } else {
        root->right = insert(root->right, value);
        if (tree_height(root->right) - tree_height(root->left) == 2) {
            if (value > root->right->value) {
                root = rotate_left(root, root->right);
            } else {
                root = rotate_right_left(root, root->right);
            }
        }
    }
    int left_height = tree_height(root->left);
    int right_height = tree_height(root->right);
    root->height = left_height > right_height?left_height+1:right_height+1;
    return root;
}

Tree delete_node(Tree root, int value) {
    if(root == NULL) {
        printf("Empty Tree.\n");
        return NULL;
    }
    if (value == root->value) {
        if(root->left == NULL) {
            PtrToNode right = root->right;
            free(root);
            root = right;
        } else if (root->right == NULL) {
            PtrToNode left = root->left;
            free(root);
            root = left;
        } else {
            PtrToNode sm_max = find_smaller_max(root);
            int sm_value = sm_max->value;
            root->value = sm_value;
            root->left = delete_node(root->left, sm_value);
            if(tree_height(root->right) - tree_height(root->left) == 2) {
                if (tree_height(root->right->right) >= tree_height(root->right->left)) {
                    root = rotate_left(root, root->right);
                } else {
                    root = rotate_right_left(root, root->right);
                }
            }
        }
    } else if (value < root->value) {
        root->left = delete_node(root->left, value);
        if(tree_height(root->right)-tree_height(root->left) == 2) {
            if(tree_height(root->right->right) >= tree_height(root->right->left)) {
                root = rotate_left(root, root->right);
            } else {
                root = rotate_right_left(root, root->right);
            }
        }
    } else {
        root->right = delete_node(root->right, value);
        if(tree_height(root->left)-tree_height(root->right) == 2) {
            if(tree_height(root->left->left) >= tree_height(root->left->right)) {
                root = rotate_right(root, root->left);
            } else {
                root = rotate_left_right(root, root->left);
            }
        }
    }
    if (root != NULL) {
        int left_height = tree_height(root->left);
        int right_height = tree_height(root->right);
        root->height = left_height > right_height?left_height+1:right_height+1;
    }
    return root;
}

PtrToNode find_smaller_max(Tree root) {
    PtrToNode iterator = root->left;
    while(iterator->right != NULL) {
        iterator = iterator->right;
    }
    return iterator;
}

Tree rotate_right(Tree root, PtrToNode node){
    root->left = node->right;
    node->right = root;
    int left_height = tree_height(root->left);
    int right_height = tree_height(root->right);
    root->height = left_height>right_height?left_height+1:right_height+1;
    left_height = tree_height(node->left);
    right_height = tree_height(node->right);
    node->height = left_height>right_height?left_height+1:right_height+1;
    return node;
}

Tree rotate_left(Tree root, PtrToNode node){
    root->right = node->left;
    node->left = root;
    int left_height = tree_height(root->left);
    int right_height = tree_height(root->right);
    root->height = left_height > right_height?left_height+1:right_height+1;
    left_height = tree_height(node->left);
    right_height = tree_height(node->right);
    node->height = left_height>right_height?left_height+1:right_height+1;
    return node;
}

Tree rotate_right_left(Tree root, PtrToNode node){
    root->right = rotate_right(node, node->left);
    return rotate_left(root, root->right);
}

Tree rotate_left_right(Tree root, PtrToNode node){
    root->left = rotate_left(node, node->right);
    return rotate_right(root, root->left);
}

void print_tree(Tree t, int type, int level) {
    if (t == NULL) {
        return;
    }
    print_tree(t->right, 2, level+1);
    switch(type) {
        case 0:
            printf("%2d\n", t->value);
            break;
        case 1:
            for(int i=0; i<level; i++) {
                printf("\t");
            }
            printf("\\ %2d\n", t->value);
            break;
        case 2:
            for(int i=0; i<level; i++) {
                printf("\t");
            }
            printf("/ %2d\n", t->value);
            break;
    }
    print_tree(t->left, 1, level+1);
}

void test_avl() {
    printf("Test AVL Tree:\n");
    printf("Commands:\n");
    printf("[0] exit\n");
    printf("[1] insert value into the tree\n");
    printf("[2] delete value from the tree\n");
    printf("[3] get height from the tree\n");
    
    Tree tree = create_tree();
    print_tree(tree, 0, 0);
    int exit = 1;
    while(exit) {
        int cmd = -1;
        scanf("%d", &cmd);
        if (cmd == 0) {
            exit = 0;
        } else if(cmd == 1) {
            int val;
            scanf("%d", &val);
            tree = insert(tree, val);
            print_tree(tree, 0, 0);
            printf("\n");
        } else if (cmd == 2){
            int val;
            scanf("%d", &val);
            tree = delete_node(tree, val);
            print_tree(tree, 0, 0);
            printf("\n");
        } else if (cmd == 3) {
            int val;
            scanf("%d", &val);
            PtrToNode t = find(tree, val);
            printf("%d\n", tree_height(t));
        }
    }
}